component/src/main/java/org/wso2/extension/siddhi/execution/streamingml/clustering/kmeans/KMeansIncrementalSPExtension.java

```diff
@@ -0,0 +1,295 @@
+                        this.attributeExpressionExecutors[1].getClass().getCanonicalName());
+            }
+        } else if (attributeExpressionExecutors[1].getReturnType() == Attribute.Type.DOUBLE) {
+            if (attributeExpressionExecutors[1] instanceof ConstantExpressionExecutor) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("decayRate is specified.");
+                }
+                decayRate = (Double) ((ConstantExpressionExecutor) attributeExpressionExecutors[1]).getValue();
+
+                if (decayRate < 0 || decayRate > 1) {
+                    throw new SiddhiAppCreationException("decayRate should be in [0,1] but given as " + decayRate);
+                }
+
+                //expressionExecutors[2] --> numberOfClusters
+                if (!(attributeExpressionExecutors[2] instanceof ConstantExpressionExecutor)) {
+                    throw new SiddhiAppCreationException("numberOfClusters has to be a constant but found " +
+                            this.attributeExpressionExecutors[2].getClass().getCanonicalName());
+                }
+                if (attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT) {
+                    coordinateStartIndex = 3;
+                    numberOfClusters = (Integer) ((ConstantExpressionExecutor)
+                            attributeExpressionExecutors[2]).getValue();
+                } else {
+                    throw new SiddhiAppCreationException("numberOfClusters should be of type int but found " +
+                            attributeExpressionExecutors[2].getReturnType());
+                }
+            } else {
+                throw new SiddhiAppCreationException("decayRate has to be a constant but found " +
+                        this.attributeExpressionExecutors[1].getClass().getCanonicalName());
+            }
+        } else {
+            throw new SiddhiAppCreationException("The second query parameter should either be decayRate or " +
+                    "numberOfClusters which should be of type double or int respectively but found " +
+                    attributeExpressionExecutors[1].getReturnType());
+        }
+
+        dimensionality = attributeExpressionExecutors.length - coordinateStartIndex;
+        coordinateValuesOfCurrentDataPoint = new double[dimensionality];
+
+        //validating all the attributes to be variables
+        for (int i = coordinateStartIndex; i < coordinateStartIndex + dimensionality; i++) {
+            if (!(this.attributeExpressionExecutors[i] instanceof VariableExpressionExecutor)) {
+                throw new SiddhiAppCreationException("The attributes should be variable but found a " +
+                        this.attributeExpressionExecutors[i].getClass().getCanonicalName());
+            }
+        }
+
+        String siddhiAppName = siddhiAppContext.getName();
+        modelName = modelName + "." + siddhiAppName;
```

Comment:
@niruhan shall we make this as a constant since its been used in several places?

Reply:
Removed in https://github.com/siddhi-io/siddhi-execution-streamingml/pull/40